mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java

```diff
@@ -162,9 +162,14 @@ public void processPublish(MqttPublishMessage msg) {
                     .thenCompose(authorized -> authorized ? doPublish(msg) : doUnauthorized(msg));
         }
         result.thenAccept(__ -> msg.release())
-              .exceptionally(ex -> {
-                    log.error("[{}] Write {} to Pulsar topic failed.", msg.variableHeader().topicName(), msg, ex);
-                    msg.release();
+                .exceptionally(ex -> {
+                    Throwable cause = ex.getCause();
+                    log.error("[Publish] [{}] Write {} to Pulsar topic failed.",
+                            msg.variableHeader().topicName(), msg, cause);
+                    // Prevent after thenAccept(__ -> msg.release()) get exceptions.
+                    if (msg.refCnt() != 0) {
```

Comment:
msg.release could not throw exception, if it throws exception, we'd better use safeRelease instead. So , here. we should not add `msg.refCnt() != 0`

Reply:
fixed